### ISSUE TITLE
Abaqus parser

### ISSUE THREAD

**Issue by dbeurle** [nschloe/meshio#493]
Thanks for the work on meshio!  Quite the task you've taken on ;-)

I'm trying to convert an Abaqus input to a gmsh file format but I am unable to get the element sets (`Elset`) and node sets (`Nset`) transferred into physical groups in the gmsh format.  I kinda need these for prescribing my boundary conditions.

What would you suggest to add in order to get these sets across?  The `Mesh` class doesn't seem to provide a mechanism for passing the element sets but does for `node_sets`.  However this does not translate to a physical group (of dimension 0) in gmsh.

Another issue:  the `node_sets` and `element_sets` do not retain the casing of the set name.  For example `applied_displacements` becomes `APPLIED_DISPLACEMENTS`, which is unexpected.  Are you happy to accept a PR to revert this behaviour?

Ideas on how to move forward?  I'm happy to do the grunt work.

**Comment by nschloe** [comment]
As far as I remember, there is no support for sets in meshio yet. See https://github.com/nschloe/meshio/issues/249 which asks about the same.

What's a set (in abaqus terms) anyway? A collection of nodes, or are the nodes stores in one big array with an additional dictionary?

**Comment by dbeurle** [comment]
They look like this:
```
*Nset, nset=applied_displacement
     2,     3,    60,    61,    62,    63,   728,   729,   786,   787,   788,   789,  1454,  1455,  1512,  1513
  1514,  1515,  2180,  2181,  2238,  2239,  2240,  2241,  2906,  2907,  2964,  2965,  2966,  2967,  3632,  3633
  3690,  3691,  3692,  3693,  4358,  4359,  4416,  4417,  4418,  4419,  5084,  5085,  5142,  5143,  5144,  5145
  5810,  5811,  5868,  5869,  5870,  5871,  6536,  6537,  6594,  6595,  6596,  6597,  7262,  7263,  7320,  7321
  7322,  7323,  7988,  7989,  8046,  8047,  8048,  8049,  8714,  8715,  8772,  8773,  8774,  8775,  9440,  9441
...
 58543,
*Elset, elset=applied_displacement
     2,     4,     6,     8,    10,   484,   486,   488,   490,   492,   966,   968,   970,   972,   974,  1448
  1450,  1452,  1454,  1456,  1930,  1932,  1934,  1936,  1938,  2412,  2414,  2416,  2418,  2420,  2894,  2896
  2898,  2900,  2902,  3376,  3378,  3380,  3382,  3384,  3858,  3860,  3862,  3864,  3866,  4340,  4342,  4344
  4346,  4348,  4822,  4824,  4826,  4828,  4830,  5304,  5306,  5308,  5310,  5312,  5786,  5788,  5790,  5792
  5794,  6268,  6270,  6272,  6274,  6276,  6750,  6752,  6754,  6756,  6758,  7232,  7234,  7236,  7238,  7240
  7714,  7716,  7718,  7720,  7722,  8196,  8198,  8200,  8202,  8204,  8678,  8680,  8682,  8684,  8686,  9160
  9162,  9164,  9166,  9168,  9642,  9644,  9646,  9648,  9650, 10124, 10126, 10128, 10130, 10132
*Nset, nset=fixed_base
    53,    54,    58,   492,   510,   511,   779,   780,   784,  1218,  1236,  1237,  1505,  1506,  1510,  1944
  1962,  1963,  2231,  2232,  2236,  2670,  2688,  2689,  2957,  2958,  2962,  3396,  3414,  3415,  3683,  3684
  3688,  4122,  4140,  4141,  4409,  4410,  4414,  4848,  4866,  4867,  5135,  5136,  5140,  5574,  5592,  5593
  5861,  5862,  5866,  6300,  6318,  6319,  6587,  6588,  6592,  7026,  7044,  7045,  7313,  7314,  7318,  7752
  7770,  7771,  8039,  8040,  8044,  8478,  8496,  8497,  8765,  8766,  8770,  9204,  9222,  9223,  9491,  9492
  9496,  9930,  9948,  9949, 10217, 10218, 10222, 10656, 10674, 10675, 10943, 10944, 10948, 11382, 11400, 11401
 11669, 11670, 11674, 12108, 12126, 12127, 12395, 12396, 12400, 12834, 12852, 12853, 13121, 13122, 13126, 13560
...
*Elset, elset=fixed_base
   459,   460,   477,   479,   481,   941,   942,   959,   961,   963,  1423,  1424,  1441,  1443,  1445,  1905
  1906,  1923,  1925,  1927,  2387,  2388,  2405,  2407,  2409,  2869,  2870,  2887,  2889,  2891,  3351,  3352
  3369,  3371,  3373,  3833,  3834,  3851,  3853,  3855,  4315,  4316,  4333,  4335,  4337,  4797,  4798,  4815
  4817,  4819,  5279,  5280,  5297,  5299,  5301,  5761,  5762,  5779,  5781,  5783,  6243,  6244,  6261,  6263
  6265,  6725,  6726,  6743,  6745,  6747,  7207,  7208,  7225,  7227,  7229,  7689,  7690,  7707,  7709,  7711
  8171,  8172,  8189,  8191,  8193,  8653,  8654,  8671,  8673,  8675,  9135,  9136,  9153,  9155,  9157,  9617
  9618,  9635,  9637,  9639, 10099, 10100, 10117, 10119, 10121, 10581, 10582, 10599, 10601, 10603
*End Part
**  
**
** ASSEMBLY
**
*Assembly, name=Assembly
**  
*Instance, name=boot-1, part=boot
*End Instance
**  
*End Assembly
```

Basically a bunch of indices that make up some kind of named object.  For the `Nset` these are node indices and for `Elset` these are element indices.  The problem is that these reference the parent element (in this case a hexahedron) but give no information about the surface structure.  They should roughly translate to `physicalNames` in gmsh.

**Comment by dbeurle** [comment]
I'm hoping these translate (roughly) to
```
$PhysicalNames
3
1 2 "fixed_base"
1 3 "applied_displacement"
3 1 "boot"
$EndPhysicalNames
```
Then I can convert these into `$Elements` with the appropriate physical tag.

What's required to get this support into meshio?  I can't be the only one that could use this functionality (apart from the previously linked issue).

**Comment by nschloe** [comment]
> What's required to get this support into meshio?

Okay, so in Abaqus we seem to have proper tags, i.e., indices into a node array. This is the way it should be imho.

gmsh, unfortuantely, handles things differently. They call it a _physical tag_, but really it's an int-valued _function_ on the elements. I'd say for reading, just dump the tags in a dictionary of `node_tags` and/or `element_tags` in the `Mesh` object. This will need a writer and a test in Abaqus, so maybe that's something to start with.

Second, you'll need a writer for such tags in gmsh. This will require checking if the tags can actually be translated into a function, i.e., check if every node appears in the tag fields once and only once etc.

**Comment by tianyikillua** [comment]
Okay, there are indeed two mathematically equivalent representations of sub-regions:
- **Using indicator functions**: to each point / cell we associate a unique integer corresponding to the sub-region. This representation is point / cell-centric and leads to its interpretation as `point_tags` or `cell_tags`. This is currently used by `gmsh`, `medit`, `flac3d` and `med` (maybe also others).
- **Using explicit enumerations of points / cells that are contained in each sub-region**. This representation is set-centric and can be named `point_sets` and `cell_sets`. In this case, maybe we can implement them by two Python dicts as follows
`
mesh.point_sets = {
    `point set A`: [0, 5, 10],
    `point set B`: [2, 7, 11],
}
mesh.cell_sets = {
    `cell set A`: {
        "triangle": [0, 2, 3],
        "quad": [1, 3, 9, 10],
    },
    `cell set B`: {
        "triangle": [1, 4, 10],
    }
}
`

With some assumptions (when some points / cells belong to multiple sets, additional care is needed), these two representations can be interchanged / converted to each other.

As for Abaqus, maybe we can begin by implementing the above `mesh.point_sets` and `mesh.cell_sets`?

**Comment by nschloe** [comment]
Thanks @tianyikillua for the write-up. Option 1 is indeed what many formats implement -- unfortunately. They think of labeling as an integer-valued function over the points/cells. (Gmsh call those functions tags, a misnomer.)

Advantages:
 * every point/cell is guaranteed to have a value

Disadvantages:
 * you can only assign one value to each cell

As opposed to that, the tag approach.

Advantages:
  * You can assign as many tags as you want to points/cells. (E.g., you have an arrange of materials, e.g., "iron", "air", etc. as well as something like "boundary")
  * Most FEM/FVM codes must loop over cell sets to produce equation systems; with tags, you can simply loop over those.

Disadvantages
  * If needed, you need to manually make sure that each point/cell has a value.

I'm definitely in favor of Option 2 (tags) even though it doesn't fit numerous mesh formats. We'll just have to convert and do some work there.

**Comment by dbeurle** [comment]
Sounds good.  Then we go with option 2 and handle the conversion.

I'll push another commit renaming `element_sets` to `cell_sets` and introduce a *new* set `point_sets`?  But how is this is different to the prexisting `node_sets` @tianyikillua?

Another small problem is the definition of part sets and instance sets, which can have the same name but the current parser considers only keywords...see the bottom of the  attached input file.  I'm not sure there is a reasonable default here.

[boot.txt](https://github.com/nschloe/meshio/files/3804451/boot.txt)

**Comment by tianyikillua** [comment]
Well, I prefer calling them `point_sets` instead of `node_sets`, and `cell_sets` instead of `element_sets`, merely for consistency reasons.

In meshio, "nodes" are called points and "elements" are called cells. Idem for point data and cell data. Different mesh formats adopt their own terminology. In Nastran for instance, nodes are termed "grids" for instance. What is really nice in meshio is the generality and the consistency across formats.

**Comment by tianyikillua** [comment]
In your file, I only see a redefinition of the point set `lower` (which is the same with or without `instance=boot-1`), idem for the cell set `lower`.

I'm not familiar with the notion of "assembly" under Abaqus. Maybe we only need to parse those that are global (without the `instance=` keyword)?

**Comment by nschloe** [comment]
> Well, I prefer calling them point_sets instead of node_sets, and cell_sets instead of element_sets, merely for consistency reasons.

+1

**Comment by dbeurle** [comment]
> > Well, I prefer calling them point_sets instead of node_sets, and cell_sets instead of element_sets, merely for consistency reasons.
> 
> +1

What consistency?  The `Mesh` container already includes `node_sets`:

```py
class Mesh:
    def __init__(
        self,
        points,
        cells,
        point_data=None,
        cell_data=None,
        field_data=None,
        node_sets=None,
```
and hence my question.  I'll update that to `point_sets` then.

**Comment by dbeurle** [comment]
> In your file, I only see a redefinition of the point set `lower` (which is the same with or without `instance=boot-1`), idem for the cell set `lower`.
> 
> I'm not familiar with the notion of "assembly" under Abaqus. Maybe we only need to parse those that are global (without the `instance=` keyword)?

It's a pickle. I choose this example because it shows that both at the part (or mesh) level a cell / point set can exist with a given name.  An instance (potentially a collection of parts) can also define the same sets with the same name (as you've seen).  A part level cell / point set will only by generated if explicitly created in the part module of Abaqus.  Any point / cell sets created as a result of applying boundary conditions (in Abaqus) will be created as assembly level cell / point sets (the latter of the two in the input file) and no part level sets will exist.

For someone only using Abaqus to generate a mesh, then only part level sets exist.  However, if comparing a model across Abaqus and another program, these will be defined only at the instance level - however the two could potentially exist at the same time without being directly related but even though they possess the same name.

Some reasonable default might need to be selected here.  How about prefer the instance level sets if it exists and fall back to the part sets if no instance is detected?

**Comment by keurfonluu** [comment]
I wonder if it would not be easier to make ``points``, ``cells``, ``point_sets`` and ``cell_sets`` as properties of ``meshio.Mesh``?

Something like:

```python
@property
def points(self):
    if self._points:
        return self._points
    else:
        # Do something with self._point_sets to get points

@points.setter
def points(self, value):
    self._points = value
```

It would avoid rewriting all the writers "just" to make Abaqus happy...

**Comment by tianyikillua** [comment]
> I wonder if it would not be easier to make `points`, `cells`, `point_sets` and `cell_sets` as properties of `meshio.Mesh`?
> 
> Something like:
> 
> ```python
> @property
> def points(self):
>     if self._points:
>         return self._points
>     else:
>         # Do something with self._point_sets to get points
> 
> @points.setter
> def points(self, value):
>     self._points = value
> ```
> 
> It would avoid rewriting all the writers "just" to make Abaqus happy...

In fact `point_sets` and `cell_sets` will simply be additional attributes to the `Mesh` class. They will be used to store / indicate subsets of points / cells.

**Comment by nschloe** [comment]
> What consistency? The Mesh container already includes node_sets:

`node_sets` should be renamed `point_sets`. I'm not sure why it's node_sets now, perhaps because of Exodus? Anyway, please change it.

**Comment by PradeepPeriyasamy** [comment]
Hi all, 

Could you please confirm if this query has been completed?
